mathics/builtin/base.py

```diff
@@ -432,6 +432,18 @@ def get_option_string(self, *params):
                     return s.get_name()[len(prefix) :], s
         return None, s
 
+    @property
+    def is_literal(self) -> bool:
```

Comment:
@rocky, I have a question: since just ``Expression`` and  ``Symbol`` are the only classes that should have a non-trivial ``evaluation`` method, wouldn't be enough to check class membership? (with ```isinstance(expr, (Symbol, Expression))```)

Furthermore, I imagined, for example, to have a superclass ``Evaluable`` from which ``Expression`` and ``[Evaluable]Symbol`` could have as a superclass (together with ``Element``).

Reply:
> @rocky, I have a question: since just `Expression` and `Symbol` are the only classes that should have a non-trivial `evaluation` method, wouldn't be enough to check class membership? (with `isinstance(expr, (Symbol, Expression))`)
> 

This is dangerous kind of thinking. Note that Boxing also has a `is_literal` as well.  Making this a property things break when the class structure changes. Using `isinstance()` things may break in a way that is silent and hard to track down. 


> Furthermore, I imagined, for example, to have a superclass `Evaluable` from which `Expression` and `[Evaluable]Symbol` could have as a superclass (together with `Element`).

First the superficial comment: I never understood or like the word Instanceable. What in plain simple dumb longer-form language does Evaluable mean. It has an evaluate function? Or something more? What does Instanceable mean in longer dumb form? 

The more fundamental worry which is really the same concern as using `isinstance()` is that basically OO is evil because it has unintended consequences. See "Inheretance vs Delegation" in https://www.cs.unc.edu/~dewan/comp401/current/Class%20Notes/22_Delegation_Notes.pdf

We sometimes use what in Ruby is called a Mixin to add particular attributes and properties. For example I did this with NumericOperators.  But I suspect we should be careful and the current thought is not to rely too much on OO strict inheritance. Maybe this Evaluable thing is a Mixin, I don't know.  Rather than describe what the mechanism is, describe what it is you want to achieve.  

The docstring for `is_literal()` tries to describe why that exists and if that's not enough I've written elsewhere why it might be useful. If there are other things, please let me know and describe what's up?

Reply:
I should also say that if `is_literal` is totally meaningless for the Builtin object, then the caller should be fixed. I wonder _why_ it is that this caller is calling "evaluate" on a builtin object that has no notion of whether it is a literal or not. 

Literalness or constantness _is_ a property of evaluation.

Reply:
> 
> First the superficial comment: I never understood or liked the word Instanceable. What in plain simple dumb longer-form language does Evaluable mean? It has an evaluate function? Or something more? What does Instanceable mean in longer dumb form?
> 

Instanceable has (at least for me) a clear meaning in Mathics: a (base) ``Builtin`` is a class that just stores information about how to build a ``Definition`` object. If you try to instanciate it, what you get is an ``Expression`` object.

 On the other hand, an ``InstanceableBuiltin`` is a kind of object that -apart from storing information about how to build a ``Definition``, can also be used as an ``Element`` in an expression. ``PatternObject`` is an example of this kind of classes.  It contains information that could be stored in an  ``Expression`` object, but the way in which the information is stored, and the methods are implemented are particular for each InstanceableBuiltin class. 



> The more fundamental worry which is really the same concern as using `isinstance()` is that basically OO is evil because it has unintended consequences because too much is implicit rather than explicit. See "Inheretance vs Delegation" in https://www.cs.unc.edu/~dewan/comp401/current/Class%20Notes/22_Delegation_Notes.pdf
>

OK, I am going to have a look
 
> We sometimes use what in Ruby is called a Mixin to add particular attributes and properties. For example I did this with NumericOperators. But I suspect we should be careful and the current thought is not to rely too much on OO strict inheritance. Maybe this Evaluable thing is a Mixin, I don't know. Rather than describe what the mechanism is, describe what it is you want to achieve.
> 

Yes, what I imagined that ``Evaluable`` could be something that supports an ``evaluate`` method, and then, an Expression (and a Symbol) would be at the time ``Element`` and ``Evaluable``. And if you define an ``InstancealeBuiltin``, you can decide if make it or not ``Evaluable``.  All the point is if you want to have an ``evaluate`` method in the base class ``Element`` or just to have them in objects on what you want to call it.




> The docstring for `is_literal()` tries to describe why that exists and if that's not enough I've written elsewhere why it might be useful. If there are other things, please let me know and describe what's up?


Yes, the docstring is clear about what this method is supposed to do. My question was about the reason for this implementation choice.

> I should also say that if is_literal is totally meaningless for the Builtin object, then the caller should be fixed. I wonder why it is that this caller is calling "evaluate" on a builtin object that has no notion of whether it is a literal or not.

a Builtin shouldn't have a ``is_literal`` because it is not an ``Element``. On the other hand, ``InstanceableBuiltin`` should have it, because they are at the time ``Builtin`` and ``Element``.

Reply:
> a Builtin shouldn't have a `is_literal` because it is not an `Element`. On the other hand, `InstanceableBuiltin` should have it, because they are at the time `Builtin` and `Element`.

Ok. Something in boxing is treating a GraphicsBox as though it were an Element: 

```
Traceback (most recent call last):
  File "mathics/docpipeline.py", line 117, in test_case
    result = evaluation.evaluate(query)
  File "github/Mathics3/mathics-core/mathics/core/evaluation.py", line 343, in evaluate
    result = run_with_timeout_and_stack(evaluate, timeout, self)
  File "github/Mathics3/mathics-core/mathics/core/evaluation.py", line 98, in run_with_timeout_and_stack
    return request()
  File "github/Mathics3/mathics-core/mathics/core/evaluation.py", line 336, in evaluate
    return self.format_output(self.last_eval, format)
  File "github/Mathics3/mathics-core/mathics/core/evaluation.py", line 429, in format_output
    result = expr.format(self, "System`OutputForm")
  File "github/Mathics3/mathics-core/mathics/core/element.py", line 305, in format
    expr = self.do_format(evaluation, form)
  File "github/Mathics3/mathics-core/mathics/core/expression.py", line 550, in do_format
    expr = super().do_format(evaluation, form)
  File "github/Mathics3/mathics-core/mathics/core/element.py", line 262, in do_format
    expr = self.create_expression(
  File "github/Mathics3/mathics-core/mathics/core/expression.py", line 1772, in _create_expression
    return Expression(head, *elements)
  File "github/Mathics3/mathics-core/mathics/core/expression.py", line 189, in __new__
    self._elements = self._build_elements(elements)
  File "github/Mathics3/mathics-core/mathics/core/expression.py", line 271, in _build_elements
    if not converted_elt.is_literal:
AttributeError: 'GraphicsBox' object has no attribute 'is_literal'
```

Reply:
Add. It to instanceblebuiltin.

Reply:
> Instanceable has (at least for me) a clear meaning in Mathics: a (base) `Builtin` is a class that just stores information about how to build a `Definition` object. If you try to instan[t]iate it, what you get is an `Expression` object.

Begin rant: 

I have no doubt that it has a meaning. In Mathics we call unescaping a string "escaping" it. And we call the nodes of a tree or an expression the "leaves".  And we name an evaluation method "apply". And we call incomplete HTML "XML".  (And I could go on with many other examples). 

Honestly I don't see the connection between Instanceable which isn't an English word and what you described in any way other than the most vague sense.

"Instanceable"  is someone trying to sound clever uses but doesn't realize they are being stupid. To the extent Instanceable has meaning it means that it is something that you are able create an instance (or object) of. Python  claims since 15 years ago everything is now an object, so that means very little: `str()` is an "instanceable" constructor because I can write `str(x)` and get back an instance of a string. 

I gather that this Instanceable thing is tied with Mathics expressions (not to be confused with the Mathics Expression class which refers to a compound expression).   Mathics expressions (derived from BaseElement) are very specific kinds of objects. 

So I kind of see how Instanceable fits in with the pattern of I'll just use a generic term like "data" and assume you know what I really am talking about. 

End rant. 

Some other time I will go over what you write and ask questions. I would like to be able to describe this in a more conventional and less vague way which seems very specific to Mathics.  That is where we probably need to start.

WL doesn't have a notion of Instanceable that I can find. Nor does it call "elements" "leaves", and so on.

Reply:
> Add. It to instanceblebuiltin.

Ok - will do when I have time. Thanks!